dotcom-rendering/src/useEffectReducer.ts

```diff
@@ -0,0 +1,79 @@
+import type { Dispatch } from 'react';
+import { useEffect, useReducer, useRef } from 'react';
+
+type StateAndEffect<State, Effect> = {
+	state: State;
+	effect?: Effect;
+};
+
+/**
+ * @param reducer Very similar to the reducer normally passed to
+ * {@linkcode React.useReducer}, but also returns a description of effects
+ * alongside the state.
+ *
+ * @param effectReducer A handler for effects. It takes a description of an
+ * effect and decides how to run that effect, returning an `Action` that will
+ * be fed back into the next loop of the `reducer`.
+ *
+ * Depending on how an application is written, different `effectReducer`s can
+ * be passed in different environments. So in production this might run actual
+ * effects, whereas in tests it might run mocked versions.
+ *
+ * *Note:* The value used will be the one passed on the initial render; the
+ * argument is ignored after this point.
+ *
+ * @param initialState The state at the start of the application; the same
+ * as the `initialState` argument usually passed to
+ * {@linkcode React.useReducer}.
+ *
+ * @param initialEffect Optionally, an effect to run at the start of the
+ * application.
+ *
+ * @returns State and a `dispatch` function, the same as
+ * {@linkcode React.useReducer}.
+ */
+const useEffectReducer = <State, Effect, Action>(
+	reducer: (
+		state: State,
+		action: Action,
+	) => { state: State; effect?: Effect },
+	effectReducer: (effect: Effect) => Action | Promise<Action>,
+	initialState: State,
+	initialEffect: Effect | undefined,
+): [State, Dispatch<Action>] => {
```

Comment:
Hmm, are you referring to the return type here? Currently this matches the pattern React tends to use for hooks. So:

```ts
const [value, setValue] = useState(initialValue);
const [isPending, startTransition] = useTransition();
const [state, dispatch] = useReducer(reducer, initialState);
const [state, dispatch] = useEffectReducer(reducer, effectReducer, initialState, initialEffect);
```

Changing it would result in an API that looks somewhat different:

```ts
const [value, setValue] = useState(initialValue);
const [isPending, startTransition] = useTransition();
const [state, dispatch] = useReducer(reducer, initialState);
const { state, dispatch } = useEffectReducer(reducer, effectReducer, initialState, initialEffect);
```

Reply:
The parameters specifically, not the return type 😝